Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -479,4 +479,164 @@ HRESULT _DWriteFontGetBoundingBoxesForGlyphs(
+            const uint8_t* underlyingBuffer = CFDataGetBytePtr(_data.get());
+            *fragmentStart = reinterpret_cast<const void*>(&(underlyingBuffer[fileOffset]));
+        }
+
+        if (fragmentContext) {
+            // Deliberately unused: this is meant to be passed to ReleaseFileFragment() below to free part of the data,
+            // but this stream frees all the underlying CFData at once
+            *fragmentContext = nullptr;
+        }
+
+        return S_OK;
+    };
+
+    void STDMETHODCALLTYPE ReleaseFileFragment(void* fragmentContext){
+        // Deliberate no-op: data is released alongside the _data member with the destruction of this object
+    };
+
+private:
+    woc::unique_cf<CFDataRef> _data;
+    uint64_t _lastWriteTime;
+};
+
+/**
+ * Custom implementation of IDWriteFontFileLoader that loads a CGDataProviderRef as its font file
+ */
+class DWriteFontBinaryDataLoader : public RuntimeClass<RuntimeClassFlags<WinRtClassicComMix>, IDWriteFontFileLoader> {
+protected:
+    InspectableClass(L"Windows.Bridge.DirectWrite", TrustLevel::BaseTrust);
+
+public:
+    DWriteFontBinaryDataLoader() {
+    }
+
+    HRESULT RuntimeClassInitialize(CGDataProviderRef dataProvider) {
+        CFRetain(dataProvider);
+        _dataProvider.reset(dataProvider);
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE CreateStreamFromKey(_In_ const void* fontFileReferenceKey,
+                                                  uint32_t fontFileReferenceKeySize,
+                                                  _Out_ IDWriteFontFileStream** fontFileStream) {
+        if (!fontFileStream) {
+            return E_INVALIDARG;
+        }
+
+        // Ignore first two params, just return the same kind of stream always
+        ComPtr<DWriteFontBinaryDataStream> ret;
+        RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataStream>(&ret, _dataProvider.get()));
+
+        *fontFileStream = ret.Detach();
+        return S_OK;
+    }
+
+private:
+    woc::unique_cf<CGDataProviderRef> _dataProvider;
+};
+
+/**
+ * Creates an IDWriteFontFace by attempting to use a CGDataProviderRef as a font file
+ */
+HRESULT _DWriteCreateFontFaceWithDataProvider(CGDataProviderRef dataProvider, IDWriteFontFace** outFontFace) {
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    ComPtr<IDWriteFontFileLoader> dwriteDataLoader;
+    RETURN_IF_FAILED(MakeAndInitialize<DWriteFontBinaryDataLoader>(&dwriteDataLoader, dataProvider));
+
+    RETURN_IF_FAILED(dwriteFactory->RegisterFontFileLoader(dwriteDataLoader.Get()));
+
+    ComPtr<IDWriteFontFile> fontFile;
+
+    // First two params are not used (see DWriteFontBinaryDataLoader::CreateStreamFromKey())
+    // However, CreateCustomFontFileReference throws E_INVALIDARG if null is passed for the first parameter
+    // Pass in a pointer to an arbitrary, unused key, to get around this.
```

Comment:
>// Pass in a pointer to an arbitrary, unused key, to get around this. [](start = 4, length = 69)

Considered passing in the CGDataProviderRef as the key here, instead of at init-time, but decided to favor type-safety instead (passing in as key would require cast-to and back-from const void*)